### PR TITLE
Keyboard shortcut to toggle block level HTML editing #9723

### PR DESCRIPTION
**Pull request by dmahendrakar**
## Description
Adds keyboard short cut for block level HTML editing - Fixes #9723

## How has this been tested?
Added snapshot test.

## Screenshots 
Screen Shot 2019-12-15 at 2 44 08 AM
Screen Shot 2019-12-15 at 2 44 56 AM
Screen Shot 2019-12-15 at 3 27 48 AM

## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

**Comment by dpkm95** [comment]
Working atop @MarkDavidTurner's stale pr https://github.com/WordPress/gutenberg/pull/11941.
Few changes from the above pr include-
- used onToggleBlockMode instead of onToggleEditor to be consistent with WordPress function naming.
- used access('m') instead on primaryAlt('v') as it's closest to global block mode toggle secondary( 'm' ). Alternate option is ctrl('m') @talldan is that a better option?

TODO: Handle reusable block case. @talldan can you point me to reference how to find out if block is a reusable type?

**Comment by dpkm95** [comment]
I found one issue when manually testing the feature. It breaks for case of group of blocks - when I try to edit the html and toggle block mode I get *block-invalid-warning*
Screen Shot 2019-12-15 at 10 39 52 AM

I don't get why the BlockCompare shows difference that I didn't introduce - I just changed text from 'para1 changed' to 'para1'. Can someone explain or point me how to debug this?

**Comment by dpkm95** [comment]
@talldan can you please help me with direction to proceed here, I think this is either a bug in existing conversion logic or I have used it wrongly.

**Comment by dpkm95** [comment]
I got input from @youknowriad that the shortcut handler should check whether the html mode is supported for the block before switching.

What is right user experience in case block doesn't support html?
- Hide the Edit as HTML shortcut in components menu?
- Show the Edit as HTML  shortcut but disable it.
- Show the Edit as HTML shortcut have the user click it and show notice in to indicate block doesn’t support html.

File: packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js [hunk 1]
```diff
@@ -124,6 +124,10 @@ const blockShortcuts = {
 			/* translators: The forward-slash character. e.g. '/'. */
 			ariaLabel: __( 'Forward-slash' ),
 		},
+		{
+			keyCombination: access( 'm' ),
+			description: __( 'Switch between Visual Editor and Code Editor.' ),
```
Comment (talldan) [inline packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js]: It'd be good to differentiate this from the other shortcut a bit more. How about something like the following:
```suggestion
			description: __( 'Switch between editing the block visually and editing its HTML.' ),
```

**Comment by youknowriad** [comment]
> What is right user experience in case block doesn't support html?

I think it's already hidden in the menu if the block doesn't support it.

**Comment by talldan** [comment]
> I got input from @youknowriad that the shortcut handler should check whether the html mode is supported for the block before switching.

@dpkm95 Thanks for taking on another issue! It looks like you have some checks in your PR already:
https://github.com/WordPress/gutenberg/pull/19152/files#diff-ad3bcfe81a33d76e53a33739c9c89b45R171

but they need to match the logic used to hide the menu item when HTML is not supported:
https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-settings-menu/block-mode-toggle.js#L16

For particular blocks like the Columns block, the `Edit as HTML` option isn't supported, so your code needs to use `hasBlockSupport` to check for that:
https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/columns/index.js#L26

It looks like it's possible to disabled HTML editing across the entire editor as well, so that should also be checked:
https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-settings-menu/block-mode-toggle.js#L39

I don't think you need to worry about changing the menu, it already hides the option when needed.

File: packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js (hunk 1 above)
Comment (dpkm95) [reply packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js]: @talldan is ctrl('m') a better option than access( 'm' )?

**Comment by dpkm95** [comment]
@talldan @youknowriad I've added the checks, now it works well.
Waiting for https://github.com/WordPress/gutenberg/pull/19100 to be merged.

**Comment by talldan** [comment]
> Meanwhile, is ctrl('m') a better option than access( 'm' )?

@dpkm95 I think that shortcut is already used by some browsers—Firefox I think uses it for muting/umuting audio, and Edge uses it to display the reading list.

**Review by talldan — APPROVED** [review]
Nice work @dpkm95! Congrats on your first contribution to the project 🎉

**Comment by talldan** [comment]
Oh, just one thing that might be useful, if you write `Fixes #9723` or `Closes #9723` in the Pull Request description, #9723 will be closed automatically as resolved when this PR is merged.

File: packages/block-editor/src/components/block-actions/index.js [hunk 2]
```diff
@@ -165,6 +167,14 @@ export default compose( [
 					innerBlocks
 				);
 			},
+			onToggleBlockMode() {
+				const block = blocks[ 0 ];
+				const isCodeEditingEnabled = select( 'core/editor' ).getEditorSettings().codeEditingEnabled;
+
+				if ( hasBlockSupport( block.name, 'html', true ) && isCodeEditingEnabled && ! isLocked && ! isReusableBlock( block ) ) {
+					toggleBlockMode( clientIds );
```
Comment (dpkm95) [inline packages/block-editor/src/components/block-actions/index.js]: @talldan should I use clientIds[0] to be consistent `const block = blocks[ 0 ];`?

**Comment by youknowriad** [comment]
now, that the keyboard shortcuts refactor landed, this probably needs a rebase.

**Comment by dpkm95** [comment]
@youknowriad sure, I'm working on it.
While going through the keyboard shortcuts refactor I realized we need to handle case where current block supports html but it's child doesn't. Do you see a possibility for such a case? Is it already handled by toggleBlockMode?

**Comment by dpkm95** [comment]
Opened another pr for the same: https://github.com/WordPress/gutenberg/pull/19329/

**Pull request closed without merging**